src/try_into_variant.rs

```diff
@@ -0,0 +1,141 @@
+use crate::utils::{AttrParams, DeriveType, State};
+use convert_case::{Case, Casing};
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+use syn::{DeriveInput, Fields, Ident, Result, Type};
+
+pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStream> {
+    let state = State::with_attr_params(
+        input,
+        trait_name,
+        quote!(),
+        String::from("try_into_variant"),
+        AttrParams {
+            enum_: vec!["ignore", "owned", "ref", "ref_mut"],
+            variant: vec!["ignore", "owned", "ref", "ref_mut"],
+            struct_: vec!["ignore"],
+            field: vec!["ignore"],
+        },
+    )?;
+    assert!(
+        state.derive_type == DeriveType::Enum,
+        "TryIntoVariant can only be derived for enums"
+    );
+
+    let enum_name = &input.ident;
+    let (imp_generics, type_generics, where_clause) = input.generics.split_for_impl();
+
+    let variant_data = state.enabled_variant_data();
+
+    let mut funcs = vec![];
+    for (variant_state, info) in
+        Iterator::zip(variant_data.variant_states.iter(), variant_data.infos)
+    {
+        let variant = variant_state.variant.unwrap();
+        let fn_name = Ident::new(
+            &format_ident!("try_into_{}", variant.ident)
+                .to_string()
+                .to_case(Case::Snake),
+            variant.ident.span(),
+        );
+        let ref_fn_name = Ident::new(
+            &format_ident!("try_into_{}_ref", variant.ident)
+                .to_string()
+                .to_case(Case::Snake),
+            variant.ident.span(),
+        );
+        let mut_fn_name = Ident::new(
+            &format_ident!("try_into_{}_mut", variant.ident)
+                .to_string()
+                .to_case(Case::Snake),
+            variant.ident.span(),
+        );
+        let variant_ident = &variant.ident;
+        let (data_pattern, ret_value, data_types) = get_field_info(&variant.fields);
+        let pattern = quote! { #enum_name :: #variant_ident #data_pattern };
+
+        let doc_owned = format!(
+            "Attempts to convert this value to the `{enum_name}::{variant_ident}` variant.\n",
+        );
+        let doc_ref = format!(
+            "Attempts to convert this reference to the `{enum_name}::{variant_ident}` variant.\n",
+        );
+        let doc_mut = format!(
+            "Attempts to convert this mutable reference to the `{enum_name}::{variant_ident}` variant.\n",
+        );
+        let doc_else = "Returns the original value if this value is of any other type.";
+        let func = quote! {
+            #[inline]
+            #[track_caller]
+            #[doc = #doc_owned]
+            #[doc = #doc_else]
+            pub fn #fn_name(self) -> Result<(#(#data_types),*), Self> {
```

Comment:
Instead of returning `Self` let's do the same thing we now do for `TryInto` where we return a custom error type: https://github.com/JelteF/derive_more/blob/master/src/convert.rs

Reply:
You might need to create a `TryIntoVariantError` type if the `TryIntoError` doesn't have the right fields.

Reply:
I'm going to create `TryIntoVariantError`.
It contains the input value and variant names of the input/output.
And it'll be displayed like `Attempted convertion to Enum::VariantB on a Enum::VariantA value`.
I'm also thinking of implementing a Deref trait to get the input field. (It seems like an antipattern tho) What do you think?

Reply:
Honestly, I'm not good at English. So please also review the documents and panic messages.